compiler/src/dotty/tools/dotc/profile/Profiler.scala

```diff
@@ -66,20 +79,59 @@ case class ProfileRange(start: ProfileSnap, end:ProfileSnap, phase:Phase, purpos
   def retainedHeapMB: Double = toMegaBytes(end.heapBytes - start.heapBytes)
 }
 
+private opaque type TracedEventId <: String = String
+private object TracedEventId:
+  def apply(stringValue: String): TracedEventId = stringValue
+  final val Empty: TracedEventId = ""
+
 sealed trait Profiler {
 
   def finished(): Unit
 
-  def beforePhase(phase: Phase): ProfileSnap
-
-  def afterPhase(phase: Phase, profileBefore: ProfileSnap): Unit
+  inline def onPhase[T](phase: Phase)(inline body: T): T =
```

Comment:
Same comment as above for all `on*` methods; we now allocate a pair on each call. I don't now if that's significant, but maybe we could just extract `(TracedEventId.Empty, Profiler.emptySnap)` and so on to constants, in order to avoid the allocations?